### PR TITLE
Update formatted-value display

### LINKED ISSUE #9546: Options for Formatted Value Display
[Issuehunt badges](https://issuehunt.io/r/directus/directus/issues/9546)

### Discussed in https://github.com/directus/directus/discussions/9040

<div type='discussions-op-text'>

<sup>Originally posted by **benhaynes** October 21, 2021</sup>
Options for...
* Enable Links — (eg: url or email) when enabled, shows a `launch` icon (18px) on hover, clicking launches link
    * External links should auto open in new tab
* Enable Copy Button — when enabled, shows a `copy` icon (18px) on hover, clicking copies the value
* Better number formatting for `9,000`, not `9000` (could use locale for formatting)
    * Maybe we can pull this from the user locale?
* Prepend Text (useful for `$123,456`)
* Appended Text (useful for `123,456 Widgets`)
* Icon After — maybe conditionally disables when using link/copy?
* Conditional Formatting Repeater
    * Operator Dropdown
    * Value Input
    * Color Chooser
    * Icon Chooser</div>

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Backers (Total: $100.00)

- [benhaynes benhaynes](https://issuehunt.io/u/benhaynes) ($50.00)
- $50.00 have been anonymously funded.
### Submitted pull Requests
- [#9773 Update formatted-value display](https://issuehunt.io/r/directus/directus/pull/9773)
---

### Tips

- Checkout the [Issuehunt explorer](https://issuehunt.io/r/directus/directus/) to discover more funded issues.
- Need some help from other developers? [Add your repositories](https://issuehunt.io/r/new) on IssueHunt to raise funds.
</details>

### PR DESCRIPTION
**Pull request by alexbrowng**
List of configuration parameters available in the display, divided into three groups, the common parameters, for all types of values, the number parameters for the types `integer,float,decimal,bigInteger`, the parameters for the types `string,text` and the conditional parameters based on the configuration of the links. Resolves #9546.

### Common parameters
- [x] **clipboard**: Adds a button to copy the field value to the clipboard.
- [x] **bold**: Change the font thickness to bold.
- [x] **font**: List of font format, allows: Sans Serif, Serif or Monospace.
- [x] **textAlign**: Text alignment list, useful to align numeric values, allows: Left, Center or Right
- [x] **color**: Text color.
- [ ] **backgroundColor**: Background color of the cell.
- [x] **iconLeft**: Icon that appears to the left of the text.
- [x] **iconLeftColor*: Color of the icon that appears to the left of the text.
- [x] **iconRight**: Icon that appears to the right of the text.
- [x] **iconRightColor**: Color of the icon that appears to the right of the text.

### Parameters for numeric types
- [x] **prefix**: Add text to the left of the value. **Important**: does not remove blanks.
- [x] **suffix**: Add text to the right of the value. **Important**: does not remove blanks.
- [ ] **multiply**: Multiply by one number. This allows, for example, to display percentage values.
- [ ] **divide**: Divide by a number. This allows, for example, to divide by the decimals of the currency when storing amounts of money without decimals, a very common case when integrating payment gateways.
- [x] **formatRules**: Allows to configure text formatting rules.
    - [x] **operator**: List of operators to check the rule, allows: Less than, Less than or equal to, Greater than, Greater than or equal to, Equal to, Not equal to, Not equal to, Not equal to.
    - [x] **value**: Value to check against the original value.
    - [x] **color**: Allows to change the color of the default text.
    - [ ] **backgroundColor**: Allows to change the default background color.
    - [x] **iconLeft**: Allows to change the icon that appears to the left of the text.
    - [x] **iconLeftColor**: Allows to change the color of the icon that appears to the left of the text.
    - [x] **iconRight**: Allows you to change the icon that appears to the right of the text.
    - [x] **iconRightColor**: Allows you to change the color of the icon that appears to the right of the text.

### Parameters for string types
- [x] **formatTitle**: Format the text.
- [x] **link**: List of link types, allows: No Link, URL, Email, Tel, WhatsApp.

### Parameters for links
- [x] **linkText**: Text to display in the link. **Important**: if no value is specified, displays the value of the default field.

### Parameters for email type links
- [x] **linkEmailSubject**: Subject of the message by default.
- [x] **linkEmailBody**: Default message body.
- [x] **linkEmailCC**: Default CC of the message.
- [x] **linkEmailBCC**: Default message BCC.

### Parameters for whatsapp type links
- [x] **linkWhatsappText**: Default message text.

---

<details>
<summary>
<b>IssueHunt Summary</b>
</summary>

### Referenced issues

This pull request has been submitted to:
- [#9546: Options for Formatted Value Display](https://issuehunt.io/repos/7122594/issues/9546)
---
</details>

### CODE REVIEW

**Comment by alejandro-tss** [comment]
## Some screenshots for numbers

### Without configuration
simple-numbers

### With configuration
formated-numbers

**Comment by benhaynes** [comment]
Whoa... my oh my! This is quite the PR!! You really went 150% on this one, @alejandro-tss ... thank you!

I'll dive into this on Monday, but in the meantime, it'd be great to see a screenshot of this Display's Options in Settings (are those link options conditional based on the dropdown? 😍 ). I love the flexibility you've included, but I want to make sure it's not overwhelming in the configuration UI. I wonder if some of these options might be a bit beyond our [80/20 rule](https://docs.directus.io/contributing/introduction/#feature-requests). 🤔 

Here are the ones I'm questioning:

* **prefix/suffix** — Any reason not to promote these to common parameters for all data types?
* **iconLeft** — (and all other left icon options) I think we can get more use out of iconRight (and just calling it "icon"). Not sure how many usecases of a left icon can't be solved by the prefix. Removing this one feature would remove four options.
* **linkWhatsappText** — I know it's a popular service, but since it's not a "standard" (eg: email), I worry about starting to include opinionated third-party platforms in Core.
* I'm on the fence about all the specific email/link options. They add value, but I'm unsure about 80/20.

Thoughts on this balance, @rijkvanzanten ?

So excited about this one! 🥳

**Comment by alejandro-tss** [comment]
@benhaynes, I haven't replied before because I was waiting to see what @rijkvanzanten thought before I started to make the changes, but, I'll answer your points:

* _prefix/suffix — Any reason not to promote these to common parameters for all data types?_ **I thought it only made sense for numeric fields, but, it doesn't cost me anything to leave it in all cases.**
* _iconLeft — (and all other left icon options) I think we can get more use out of iconRight (and just calling it "icon"). Not sure how many usecases of a left icon can't be solved by the prefix. Removing this one feature would remove four options._ **You are right that there are many elements, I thought to put them in a group that was collapsible, but, it seems to me that forcing them to be in a place is a very rigid option, I gave the possibility to choose where to show the icons because, I understand that the numbers in a table should appear aligned to the right and perhaps it makes more sense to add an icon to the left or vice versa, in any case, it seemed more interesting for the user to decide.**
* _linkWhatsappText — I know it's a popular service, but since it's not a "standard" (eg: email), I worry about starting to include opinionated third-party platforms in Core._ **It seems to me well, we can remove it, this can be solved with specific extensions when the marketplace appears.**
* _I'm on the fence about all the specific email/link options. They add value, but I'm unsure about 80/20._  **In this case, I think that if you use email links, it is not superfluous to indicate default values, taking into account that they are fields that only appear when selecting that type of link and are optional.**


Finally, regarding the two fields multiply and divide, what is your opinion? I have some use cases in which it would be very useful, but, I wanted to know your opinion before implementing it.

I will record a short video to show you the configuration options, I will upload it soon.

**Comment by alejandro-tss** [comment]
@benhaynes here is a sample configuration: https://www.loom.com/share/c4aead6fc636447f8bf26cf774161505

**Comment by benhaynes** [comment]
Yeah, @rijkvanzanten will be back on Core soon, but has been 100% on our new Cloud for a while now.

Let's make prefix/suffix global, remove the left icon (and rename the right one to just "icon") and remove WhatsApp. We can leave the link stuff in for now, and see what Rijk thinks.

For multiply/divide, I thought it was overkill at first, but then was agreeing that it might be nice for percentages. Again, let's see what @rijkvanzanten thinks on this one.

Lastly, do we have the formatting rule repeater in? I noticed it's checked off but didn't see it in the video. 🤔 

Thank you!! It's looking great!! I'll come up with the Display Option order once we have the options themselves finalized.

**Comment by Fusseldieb** [comment]
Oh wow! This looks lit ahahaha 🔥

File: app/src/displays/formatted-value/index.ts [hunk 1]
```diff
@@ -8,62 +8,414 @@ export default defineDisplay({
...
+				meta: {
+					interface: 'select-dropdown',
+					width: 'half',
+					options: {
+						choices: [
+							{ text: '$t:displays.formatted-value.no_link', value: '' },
+							{ text: '$t:displays.formatted-value.link_url', value: 'url' },
+							{ text: '$t:displays.formatted-value.link_email', value: 'email' },
+							{ text: '$t:displays.formatted-value.link_tel', value: 'tel' },
+							{ text: '$t:displays.formatted-value.link_whatsapp', value: 'whatsapp' },
```
Comment (rijkvanzanten) [inline app/src/displays/formatted-value/index.ts:294]: lets remove this

**Review by rijkvanzanten — CHANGES_REQUESTED** [review]

File: app/src/displays/formatted-value/formatted-value.vue [hunk 2]
```diff
@@ -1,54 +1,205 @@
...
+			default: '',
+		},
+		linkText: {
+			type: String,
+			default: '',
+		},
+		linkTarget: {
+			type: String,
+			default: '_blank',
+		},
```
[the remark this answers is not in the document]
Comment (alejandro-tss) [reply app/src/displays/formatted-value/formatted-value.vue:110]: How do you want me to check if the link is out of the application, before or after calculating the href?

[the remark this answers is not in the document]
Comment (rijkvanzanten) [reply app/src/displays/formatted-value/formatted-value.vue:90]: If the whole href is passed in, we can check if it starts with or without a protocol 👍🏻

**Comment by alejandro-tss** [comment]
@rijkvanzanten I have resolved all change requests

**Comment by joselcvarela** [comment]
Thank you @alejandro-tss 
I only partially tested it, but I found a weird behaviour.

https://user-images.githubusercontent.com/14039341/144760755-3c170278-bff0-4da1-8884-0e96935a4825.mov

It seems you are modifying the values after they were populated or something like that.
If possible, can you fix that?

**Comment by alejandro-tss** [comment]
@joselcvarela I'm not sure I understand the error, do you mean that when editing the display settings the indicator appears that the value has been modified? If so, I think that is the default behavior, I have tried on other displays and they work the same, it seems that if the current value is different than the default value that indicator appears, sorry if I am not understanding the problem.

**Comment by joselcvarela** [comment]
Yeah that was what I intended to describe.
Just checked the main branch also behaves the same. Maybe we can improve that on other PR. 
Until the end of the day I will take another look, but seems great so far 👍

**Review by joselcvarela — APPROVED** [review]
I made a few commits to fix minor issues.
But, I also applied the conditions to strings too. I don't find a reason to limit only to numbers and this way we can be more versatile.

About the translations, @rijkvanzanten should we take advantage of existent translations or it is right like how it is done?

**Comment by rijkvanzanten** [comment]
> About the translations, @rijkvanzanten should we take advantage of existent translations or it is right like how it is done?

It's fine for now. Ideally, we figure out where else the same strings are used, so we can prevent duplicates, but that's a bit of a bigger cleanup operation not necessarily related to this new feature.

**Comment by benhaynes** [comment]
Hey @alejandro-tss — we are about to merge this, but noticed a few of my last points weren't finished. Once these are addressed I think we're good:

* Icon Right (and all accompanying options) should be removed
* Icon color should be removed; icon color should match the text color
* Icon Left should be renamed: `icon`
* Is background color still missing? I see it in one of the videos...
* Update the Options Layout to:

```
Format + Copy
Bold + Italic
Font + Align
Prefix + Suffix
Color + Background
Icon
Conditional Formatting (Full Width)
* Operator + Value (these are the conditions)
* Color + Background
* Text + Icon (text would override the text displayed)
----
Type Specific Options
```

Let me know if you want to discuss!!

**Comment by alejandro-tss** [comment]
I've been making some adjustments regarding your last comment, but, I think at some point we have taken different paths and you have to tell me where to continue, these are some things that I don't know how to solve.

At @rijkvanzanten's request I removed the following options:

- copy: set it to layout level
- background: cannot be applied at cell level.
- align: cannot be applied at cell level.

To note:

- format: this is a conditional option, only appears if the field type is string or text.

The applied changes are:

- I have removed the iconRight, iconRightColor and iconLeftColor.
- I have changed the iconLeft option to icon and the same color as the text is applied.

Therefore the options are as follows: 

Bold + Italic
Font + Color
Prefix + Suffix
Icon [+ Format]
Link [+ LinkTemplate]

Conditional Formatting (Full Width)
Name 
Filter (Full Width)
Color + Icon

Finally, I don't understand the behavior of the conditional text field, could you give me more details?

**Comment by benhaynes** [comment]
Hey @alejandro-tss, sorry for not posting an update yet, but we just had an internal meeting about this a few hours ago. @joselcvarela can give you an update (GitHub or Discord) for what the strategy is, and/or help get this wrapped up.

**Comment by joselcvarela** [comment]
Hello @alejandro-tss 

So this is the UI for the display configuration we want:
```
Format + Font
Bold + Italic
Prefix + Suffix
Color + Background
Icon + Border
Conditional Formatting (Full Width)
* Operator + Value (these are the conditions)
* Color + Background
* Text + Icon (text would override the text displayed)
``` 

Description

| Field      	| Label       	| Notes                                                                                                   	|
|------------	|-------------	|---------------------------------------------------------------------------------------------------------	|
| Format     	| Auto format 	| In case of string, uses auto format case. In case of number, adds , or . to numbers depending on locale 	|
| Font       	| Implemented 	|                                                                                                         	|
| Bold       	| Implemented 	|                                                                                                         	|
| Prefix     	| Implemented 	|                                                                                                         	|
| Suffix     	| Implemented 	|                                                                                                         	|
| Color      	| Color       	| Color for text, icon and border                                                                         	|
| Background 	| Background  	| Color for background. Should be rounded see example on image 1                                          	|
| Icon       	| Icon        	| Should be applied to the right of text                                                                  	|
| Border     	| Border      	| Boolean (by default we use 2px). Should                                                                 	|
| Conditions 	|             	|                                                                                                         	|
| Operator   	| Operator    	| Use the same component as shown on image 2                                                              	|
| Value      	| Value       	| Use the same component as shown on image 2                                                              	|
| Color      	| Color       	| Will override the color, if rule matches                                                                	|
| Background 	| Background  	| Will override the background, if rule matches                                                           	|
| Text       	| Text        	| Will override the text / value, if rule matches.                                                        	|
| Icon       	| Icon        	| Will override the icon, if rule matches.                                                                	|

Image 1
image

Image 2
image


About Image 2, you can find that component here: https://github.com/directus/directus/blob/ffde89d7e2ff7f21f5d7bcb561dc301e8208658d/app/src/panels/metric/index.ts#L182 (It is being used in Insights > Panel > Metric)

**Comment by alejandro-tss** [comment]
@joselcvarela I understand that now the conditional formatting reappears only when the field type is numeric, right?

**Comment by benhaynes** [comment]
No, it would work for strings too — just with limited operators (eq only?).

**Comment by alejandro-tss** [comment]
I do whatever you tell me, I can include more operators for strings besides eq, such as startswith, endswith or contains, as well as whether they are case sensitive or not. @benhaynes, @rijkvanzanten, @joselcvarela what do you think?

**Comment by alejandro-tss** [comment]
I have one more doubt, with the text field of the conditions, what is the behavior, does it replace all prefix + value + suffix by text or is it a replacement only of the original value, remaining prefix + text + suffix?

**Comment by benhaynes** [comment]
I think those operators are all good... though additional options for case insensitive might be too much. For the conditional text, let's have it replace prefix+value+suffix... so you could have `Free!` instead of `$Free!`.

**Comment by alejandro-tss** [comment]
One more, just to clarify, the icon should appear on the right, right? I say this because in the example image that @joselcvarela has posted the icon appears on the left.

**Comment by benhaynes** [comment]
Haha, good question! I think let's start with the icon on the **left**... that's where I see it most often.

**Comment by alejandro-tss** [comment]
I have just added a commit with the last requested changes, I am waiting for your feedback. @benhaynes, @rijkvanzanten, @joselcvarela

**Comment by benhaynes** [comment]
Awesome, thank you!  We'll take a look as soon as we can...

**Comment by joselcvarela** [comment]
Thank you @alejandro-tss 
Seems great now 👍

@benhaynes once you don't like "pulling" 😅 here you go:

Number:
https://user-images.githubusercontent.com/14039341/145764867-85818b5a-756a-4a7e-b0dc-fb3743e68179.mov

String:
https://user-images.githubusercontent.com/14039341/145764889-37bf3315-6775-4b50-815c-3950fb980431.mov

**Comment by benhaynes** [comment]
❤️  Haha, thanks @joselcvarela!! Yup, lazy... but really appreciate these videos. This looks perfect to me, hopefully we can get it merged in! Thank you, @alejandro-tss!!

CC @rijkvanzanten

**Review by rijkvanzanten — APPROVED** [review]

**Comment by neo7-studio-web** [comment]
Hi, what a great job, thanks !

Maybe I missed something, but I couldn't find any option to make a link from a string field in 9.3.0.
How does it work ?

**Comment by joselcvarela** [comment]
Hello @neo7-studio-web 
We have decided that link should have their own display.
So, this is not available as of right now.

**Comment by nervusdm** [comment]
How do I make the value use a column from the same collection ?
In your example I see 9000.
Can I set {{other_field}} ?

**Pull request merged**